07_errors/n0npax/leveldbStore.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"encoding/json"
+	"strconv"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+var levelDBPath = "/tmp/leveldb"
+
+// NewLevelDBStorer creates new storer for leveldb
+func NewLevelDBStorer() (*LevelDBStore, error) {
+	db, err := leveldb.OpenFile(levelDBPath, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &LevelDBStore{nextID: 0, ldb: db}, nil
+}
+
+// CreatePuppy creates puppy
+func (l *LevelDBStore) CreatePuppy(p *Puppy) (int, error) {
+	l.Lock()
+	defer l.Unlock()
+	p.ID = l.nextID
+	id, err := l.putPuppy(l.nextID, p)
+	l.nextID++
+	return id, err
+}
+
+// ReadPuppy reads puppy from backend
+func (l *LevelDBStore) ReadPuppy(id int) (*Puppy, error) {
+	byteID := []byte(strconv.Itoa(id))
+	if puppyData, err := l.ldb.Get(byteID, nil); err == nil {
+		var puppy Puppy
+		err := json.Unmarshal(puppyData, &puppy)
+		if err != nil {
+			return nil, Errorf(ErrCodeInternal, "could not cast stored data to puppy object")
+		}
+		return &puppy, nil
+	}
+	return nil, Errorf(ErrCodeNotFound, "iuppy with ID (%v) not found", id)
+}
+
+// UpdatePuppy updates puppy
+func (l *LevelDBStore) UpdatePuppy(id int, puppy *Puppy) error {
+	if id != puppy.ID {
+		return Errorf(ErrCodeInvalidInput, "ID is corrupted. Please ensure object ID matched provided ID")
+	}
+	l.Lock()
+	defer l.Unlock()
+	if _, err := l.ReadPuppy(id); err != nil {
+		return Errorf(ErrCodeNotFound, "puppy with ID (%v) not found", id)
+	}
+	_, err := l.putPuppy(id, puppy)
+	return err
+}
+
+// DeletePuppy deletes puppy
+func (l *LevelDBStore) DeletePuppy(id int) (bool, error) {
+	l.Lock()
+	defer l.Unlock()
+	if _, err := l.ReadPuppy(id); err != nil {
+		return false, Errorf(ErrCodeNotFound, "puppy with ID (%v) not found", id)
+	}
+	byteID := []byte(strconv.Itoa(id))
+	err := l.ldb.Delete(byteID, nil)
+	dbErrorPanic(err)
```

Comment:
You're still panicking across an API boundary here. This is easy to fix because you already return an error.